### PR TITLE
bluetooth: host: Fix broken directed advertising

### PR DESCRIPTION
**Pull request by pdunaj**
This code probably got broken during some merge.

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by pdunaj** [comment]
I have check this on PCA20041 and device seems to send correct direct advertising packets.

Without the fix code crashes.

**Comment by pdunaj** [comment]
@kapi-no , @cvinayak , @sjanc , please have a look

File: subsys/bluetooth/host/hci_core.c
Comment (cvinayak) [inline subsys/bluetooth/host/hci_core.c]: This has to remain CONFIG_BT_CENTRAL, as auto connect state in central role needs the scanner started again.

File: subsys/bluetooth/host/hci_core.c
Comment (cvinayak) [inline subsys/bluetooth/host/hci_core.c]: This code should not be in central code path.

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Review by sjanc — CHANGES_REQUESTED** [review]
I think this should be in form of:

if (err) {
   if (IS_ENABLED(PERIPHERAL) {
 // handle peripheral only error
 } 

if (IS_ENABLED(CENTRAL) {
  //handle central only errors
}
}


or maybe i switch-case... whatever makes code easier to read

**Comment by pdunaj** [comment]
@cvinayak , @sjanc could you re-review?

**Comment by pdunaj** [comment]
ping

**Comment by jhedberg** [comment]
@pdunaj looks generally good to me, but you'll need to fix the Gitlint issue before this can be merged. I think that's because of you having a "subsys:" prefix, which is not allowed. Bluetooth commits should start with "Bluetooth: ..." (just check the commit history to see that this is the consistent form).

**Comment by pdunaj** [comment]
@cvinayak , @sjanc , can you recheck?

File: subsys/bluetooth/host/hci_core.c
Comment (cvinayak) [inline subsys/bluetooth/host/hci_core.c]: Prefer checking for relevant  HCI error code here.

File: subsys/bluetooth/host/hci_core.c
Comment (cvinayak) [inline subsys/bluetooth/host/hci_core.c]: Is it ok to switch states back-to-back from BT_CONN_CONNECT_SCAN to BT_CONN_DISCONNECTED?

**Review by cvinayak — DISMISSED** [review]

File: subsys/bluetooth/host/hci_core.c
Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: can you tell what you mean?

File: subsys/bluetooth/host/hci_core.c
Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: how can I know? :)
CO opinion is needed

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: It seems that you shouldn't be doing it, i.e. leave it in CONNECT_SCAN if that was set. It's e.g. what commit 861bae22b80 was trying to fix. Based on that commit it would also be ok to first transition to DISCONNECTED and after that to CONNECT_SCAN. @sjanc it'd be good to get your review of this detail as well since you're the author of the commit in question.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c
Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: state set removed

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: @pdunaj the state change to DISCONNECTED still needs to be there though. I'm not sure if it needs to be there first always, or only for the case where we don't go to CONNECT_SCAN. However, if you want to be in line with what the code used to do, then in case of a non-zero status the state should always first go to to DISCONNECTEd. @sjanc 	did you have a chance to look at this yet?

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: Also, it'd be nice if someone actually tested this in practice so we're not just talking theoretical stuff here :)

**Review by sjanc — CHANGES_REQUESTED** [review]
we must go through disconnected state to notify application about failed connection attempt

since only 2 error codes are allowed and both are limited to specific roles, I suggest doing something like this

	if (evt->status) {
		if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
			/*
			 * Handle advertising timeout after high duty directed
			 * advertising.
			 */
			if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
				atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
				goto done;
			}
		}

		if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
			/* handle cancelled of outgoing connection attempt */
			if (conn->err == BT_HCI_ERR_UNKNOWN_CONN_ID) {
				/* we notify before checking autoconnect flag
				 * as application may choose to change it from
				 * callback
				 */
				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);

				/* check if device is market for auto connect */
				if (atomic_test_bit(conn->flags, BT_CONN_AUTO_CONNECT)) {
					bt_conn_set_state(conn, BT_CONN_CONNECT_SCAN);
				}
				goto done;
			}
		}

		/* no other errors are allowed for this event */
		BT_ASSERT(false);
	}

**Comment by pdunaj** [comment]
hi @sjanc , thanks. Thank snippet helps a lot. I am not expert in these things so I appreciate direct help :)

Please check the new code. On top to what you proposed I added check for actual role set. I assume some device may have both central and peripheral enabled and we need to distinguish behavior based on role too. Please correct me if I am wrong.

**Comment by pdunaj** [comment]
@sjanc , I have not tested code yet btw. When I do it I will write it explicitly so for every body - do not merge this yet.

File: subsys/bluetooth/host/hci_core.c
Comment (sjanc) [inline subsys/bluetooth/host/hci_core.c]: you cannot access other fields if status is not success...

error code identifies connection role in this case

**Review by sjanc — CHANGES_REQUESTED** [review]

**Comment by pdunaj** [comment]
@sjanc , I have tested the code and it seems working. I was able to connect to central (both devices using Zephyr). Peripheral does not crash if central is not present. Central does not crash if peripheral is not present. I guess this does not cover all code paths but at basics seem ok.

**Comment by pdunaj** [comment]
@cvinayak , @jhedberg , @sjanc , can you re-review and green tick this today so we will have it off our heads :)

File: subsys/bluetooth/host/hci_core.c
Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: removed check for role

**Review by sjanc — APPROVED** [review]
looks good to me

File: subsys/bluetooth/host/hci_core.c
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: This has me a little bit nervous. Are we really sure that there's no other status codes that we can get, i.e. this has been double-checked with the HCI spec.

Also, we generally reserve asserts only for things which indicate a bug in *our* code. I.e. buggy behavior in an external component (the controller in this case) would only result in an error or ignoring the data on our side.

Unrelated to the above (and this PR) it'd be nice to have a BT_ASSERT_NOT_REACHED() that'd do the same as BT_ASSERT(false) but be much clearer in its intent for the casual reader.

Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: @sjanc , any idea?

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: @pdunaj sorry to keep dragging this PR on - I also want to see it merged ASAP, however this is really something that'd be good to sort out first. What I'd suggest is to get rid of the assert and instead have the following kind of branches in two places for the specific errors, but let the code otherwise continue executing even if we get an unexpected error:
```
if (evt->status != SOME_EXPECTED_ERROR) {
        BT_WARN("Got unexpected status 0x%02x", evt->status);
}
```

Comment (sjanc) [reply subsys/bluetooth/host/hci_core.c]: I don't think w can get other error codes but yeah, buggy controller could trigger it. So for other error code we could warn and just drop conn. I initially added assert since in case of failure error code indicate which connection (slave or master) failed. So for other error we don't know whats the expected outcome (eg was advertising stopped? etc)

Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: Warning added instead of assert.

File: subsys/bluetooth/host/hci_core.c
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: Won't this now potentially leave a connection in "connecting" state even though we got a connection complete with error status? That isn't much better than an assert, since it may leave the device non-functional. For the central branch I'd have logged the error `if (conn->err != BT_HCI_ERR_UNKNOWN_CONN_ID)` but otherwise done the same steps for the connection to clean up the state.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c
Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: My point is, even if we get a unexpected error, the one thing we do know is that the connection is disconnected, so leaving it in any other state is probably not a good idea. @sjanc btw, where in the core spec does it say that we can't trust the bdaddr if we get a non-zero status?

Comment (pdunaj) [reply subsys/bluetooth/host/hci_core.c]: Considering the old code was merge mess having the known issues handled is a huge improvement. 
Personally I would stick with assert (note that original code was using `BT_ASSERT(evt->status == 0)`) especially if no known errors come to your mind.

Anyway I don't feel competent to add any more changes. Please provide a snippet that I should put and please agree among yourselves what should be there. When it is ready and you reached an agreement please call me.

Comment (sjanc) [reply subsys/bluetooth/host/hci_core.c]: @jhedberg iit is described in Vol 2, Part E, 4.5 "Command Error Handling"

If we get other status code we won't be able to determine which connection failed to establish anyway in case both roles are pending (which leads to another issue, we should not use find_pending_connect() but role dependent function for locating pending conn).

So to avoid assert we can report error but host will probably be left in broken state whatever we do...

**Comment by jhedberg** [comment]
I'm just going to merge this now, since it either way improves the current state of the tree. We can then consider follow-up patches for further improvements.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
